cai_causal_graph/time_series_causal_graph.py

```diff
@@ -289,36 +289,56 @@ def get_minimal_graph(self) -> TimeSeriesCausalGraph:
 
     def _get_time_topological_order(self, ordered_nodes: List[str]) -> List[str]:
         """Return the provided list if it is ordered in time; otherwise, an empty list is returned."""
-
         # Iterate through the list to ensure it respects the time ordering.
         for j, node in enumerate(ordered_nodes):
             if j == 0:
                 continue
             assert isinstance(node, str)  # for linting
             # check if the time delta is correct
-            if self.get_node(ordered_nodes[j - 1]).time_lag > self.get_node(ordered_nodes[j]).time_lag:   # type: ignore
+            if self.get_node(ordered_nodes[j - 1]).time_lag > self.get_node(ordered_nodes[j]).time_lag:  # type: ignore
                 return []
 
         return ordered_nodes
 
-    def get_topological_order(self, return_all: bool = False) -> Union[List[str], List[List[str]]]:
+    def get_topological_order(
+        self, return_all: bool = False, respect_time_ordering: bool = True
+    ) -> Union[List[str], List[List[str]]]:
         """
         Return the topological order of the graph that is ordered in time.
 
+        :param return_all: If `True`, return all the possible topological orders. Default is `False`.
+        :param respect_time_ordering: If `True`, return the topological order that is ordered in time. Default is
+            `True`.
+
         For more details, see `cai_causal_graph.causal_graph.CausalGraph.get_topological_order`.
         """
-        ordered_nodes = super().get_topological_order(return_all=True)
-        top_order_list: List[List[str]] = []
-        for current_top_order in ordered_nodes:
-            assert isinstance(current_top_order, list)  # for linting
-            new_order = self._get_time_topological_order(current_top_order)
-            if new_order not in top_order_list and len(new_order) > 0:
-                top_order_list.append(new_order)
-
-        if not return_all:
-            top_order_list = top_order_list[0]  # type: ignore
-
-        return top_order_list
+        if respect_time_ordering:
+            if return_all:
+                ordered_nodes_list = super().get_topological_order(return_all=return_all)
+                # ordered_nodes_list must be a list of lists of strings
+                assert isinstance(ordered_nodes_list, list)
+                assert all(isinstance(x, list) for x in ordered_nodes_list)
+
+                ordered_nodes: List[List[str]] = []
+
+                for i in range(len(ordered_nodes_list)):
+                    assert isinstance(ordered_nodes_list[i], list)
+                    tmp = self._get_time_topological_order(ordered_nodes_list[i])  # type: ignore
+                    if tmp not in ordered_nodes and len(tmp) > 0:
```

Comment:
I don't really understand why you are checking `if tmp not in ordered_nodes`? Why is it not sufficient to just check `if len(tmp) > 0`?

As far as I understand, the `ordered_nodes_list` contains unique topological orders. The `self._get_time_topological_order` will either return the input order unchanged (if it respects time) or return an empty list. Hence, if a non empty list is returned, then it must be unique.

I think checking `if tmp not in ordered_nodes` may become somewhat computationally expensive (relatively speaking) when you have a huge graph with a lot of different topological sorts.

Reply:
There could be duplicates. I remember it from the previous implementation. But, for the sake of efficiency we could remove the check and leave duplicates since it's not conceptually wrong

Reply:
Happy to remove the check

Reply:
Well thing is if there are duplicates, then you should also be removing them below when `respect_time_order = False`. Otherwise, it isn't consistent?

I don't mind (and have approved the PR), just seemed off to me. I would say - either you should check for duplicates in all cases, or not check for them at all.

Reply:
I would not return any duplicates. The list of valid top sorts should be unique.

Reply:
Sorry, I was confusing the old implementation where I was sorting. Here we filter so no need to check.